### PR TITLE
Make go test ./... succeed by default

### PR DESCRIPTION
**Pull request by twpayne**
This PR makes

```console
$ go test ./...
```

succeed by default. Before, it would fail due to many unit tests actually being integration tests needing a running etcd server and the requirement for various linker flags to be present to set various test configuration variables.

See the individual commits for more details.

### CODE REVIEW

**Comment by twpayne** [comment]
Oh, @maintainer-s-little-helper, commit 98e2a21 is no longer part of this PR, so I'm not sure why you're complaining about it.

**Comment by twpayne** [comment]
test-me-please

**Review by tklauser — CHANGES_REQUESTED** [review]
There seems to be a mixture of `integration_test` (singular) and `integration_tests` (plural) used in the Go file build tags. I assume we want to consistently use `integration_tests` given that this is what's set in the `unit-tests` target in the `Makefile`.

**Comment by twpayne** [comment]
Thanks @tklauser, fixed.

File: pkg/datapath/loader/datapath_sha256.go
Comment (joestringer) [inline pkg/datapath/loader/datapath_sha256.go]: This file is autogenerated, right? Should we avoid committing it to the tree?

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (joestringer) [inline pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: Please add the SPDX + copyright lines here.

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (joestringer) [inline pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: We also have the option to add this program under `tools/` rather than `pkg/datapath/loader/internal/...`, I think that'd be an easier place to find. At the same time this is likely a tool we write once and forget about, so I'm not too fussed if you prefer it to stay here.

File: Makefile
Comment (joestringer) [inline Makefile]: I *think* there might be a problem with parallel builds here. The `build` target says that it depends on the datapath sha and various subdirectories. So `make -jN` would try to build each of these dependencies independently, but if the scheduling of the datapath sha isn't quick enough then plausibly the daemon code would be compiled before the file is available (or updated), and the build could either fail (no such file) or encode the wrong datapath sha.

I think this file should be a dependency of the agent itself under `daemon/Makefile`?

File: Makefile
Comment (joestringer) [inline Makefile]: I agree that some of these tests are integration tests, and tagging them as such is a good step forward.

What I like about this: Calling things out explicitly.

What I don't like about this particular line: Now we're explicitly defining the target `unit-tests` as integration tests, even though the majority of the tests are actually unit tests.

I'd like to see these to be part of a separate make target if we're going to call them out separately - maybe something like `make integration-tests` or `make go-integration-tests`?

At the same time, I don't want to lose coverage over these so if these particular tests are being run via a different target then let's make sure to grep around for where the `unit-tests` target is used, and update those locations to also call this target.

If we split this out, we can also move the `start-kvstores` / `stop-kvstores` logic over there, which will speed up this target. We can also update the docs so that we're not telling developers "to run the integration tests, run `make unit-tests`."

File: pkg/endpoint/endpoint_status_test.go
Comment (joestringer) [inline pkg/endpoint/endpoint_status_test.go:4]: I don't think that most of these `pkg/endpoint` tests are actually integration tests.

I believe that due to tech debt around the code structure, they introduce dependencies on the BPF subsystem, but that's more of a code / compile issue rather than being relevant to the actual test execution.

The right solution is to deal with that technical debt. That's been a desirable item for a long time though, and it's not at all reasonable to require this from you in this PR. A middle ground might be to use a different tag, rather than `integration_tests` we have something more apt. But that probably makes it a bit more of a hassle to manage running those tests via the make targets... so if that's still pretty painful then I'm willing to accept this.

**Review by joestringer — CHANGES_REQUESTED** [review]
Thanks for tackling this. I think there's just a couple of fairly minor bits I'd like to see changed but overall this seems like it will have a positive impact on the workflow both for general go familiarity, and also for developers on other platforms like osx who cannot run some of these more integration-style tests.

File: pkg/endpoint/endpoint_status_test.go
Comment (joestringer) [reply pkg/endpoint/endpoint_status_test.go:4]: (I didn't look _too_ closely at these so I could also be off in my assessment here. If the comment seems off-base then feel free to ignore it).

File: pkg/datapath/loader/loader_test.go
Comment (kkourt) [inline pkg/datapath/loader/loader_test.go:54]: Will this work regardless of where `go test` is invoked from?

That is, will it work both when running `go test ./...`  from the top dir,  `go test` inside `pkg/datapath/loader`, or any other option?

Comment (kkourt) [inline pkg/datapath/loader/loader_test.go:43]: Will this work regardless of where go test is invoked from?

That is, will it work both when running `go test ./...` from the top dir, `go test` inside pkg/datapath/loader, or any other option?

**Review by kkourt — APPROVED** [review]
Thanks!

LG from my side, assuming that invoking `go test` from any subdir would properly work (e.g., see comment) and that other concerns are also addressed.

File: pkg/datapath/loader/datapath_sha256.go
Comment (tklauser) [reply pkg/datapath/loader/datapath_sha256.go]: In case we decide to add it to the tree, let's add a comment header saying:

```go
// Code generated by $(cmd). DO NOT EDIT
```

So it will be recognized by tooling and e.g., GitHub as auto-generated. Also see https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (tklauser) [inline pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: As noted above on Joe's review, let's add a comment header saying:

```go
// Code generated by $(cmd). DO NOT EDIT
```

so the will be recognized by humans and tools as auto-generated. Also see https://pkg.go.dev/cmd/go#hdr-Generate_Go_files_by_processing_source

File: pkg/elf/elf_test.go
Comment (nebril) [inline pkg/elf/elf_test.go:26]: Same question as kkourt's above - will this work when run from within `pkg/elf`?

File: Documentation/contributing/testing/unit.rst
Comment (nebril) [inline Documentation/contributing/testing/unit.rst:31]: I know that this PR is a step forward in terms of naming the test suites correctly, but having `integration` here followed by `make unit-tests` would scream typo for me when reading the docs.

Would renaming the make target be possible, or is this out of scope for this PR?

Maybe we can create a `integration-tests` target as an alias for `unit-tests` so that the docs look good and we maintain backwards compatibility?

File: pkg/datapath/loader/datapath_sha256.go
Comment (twpayne) [reply pkg/datapath/loader/datapath_sha256.go]: For compatibility with the standard Go tooling we need to commit generated files to the tree. The `go` tool does not run any extra build steps by default, so if we don't commit the generated file, anyone checking out the code and running `go build` or `go test` will get an error about `DatapathSHA256` being undefined.

I've added the "Code generated..." header.

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (twpayne) [reply pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: Done.

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (twpayne) [reply pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: Moved to `tools/`.

File: Makefile
Comment (twpayne) [reply Makefile]: Thanks, done.

File: Makefile
Comment (twpayne) [reply Makefile]: OK, I've renamed the `unit-tests` target to `integration-tests` and updated the documentation, which should resolve this.

File: pkg/endpoint/endpoint_status_test.go
Comment (twpayne) [reply pkg/endpoint/endpoint_status_test.go:4]: OK if I fix the endpoint tests in a follow-up PR?

File: pkg/datapath/loader/loader_test.go
Comment (twpayne) [reply pkg/datapath/loader/loader_test.go:43]: Yes, this will work. For any Go test, the working directory is the directory of the package (i.e. where the `_test.go` file is) so these relative paths work.

File: pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go
Comment (twpayne) [reply pkg/datapath/loader/internal/generate-cilium-datapath-sha256/main.go]: Done.

File: pkg/elf/elf_test.go
Comment (twpayne) [reply pkg/elf/elf_test.go:26]: Yes.

File: Documentation/contributing/testing/unit.rst
Comment (twpayne) [reply Documentation/contributing/testing/unit.rst:31]: Done. I've renamed the `unit-tests` target to `integration-tests`.

**Comment by twpayne** [comment]
test-me-please

**Comment by twpayne** [comment]
test-me-please

**Comment by jrajahalme** [comment]
> Done. I've renamed the `unit-tests` target to `integration-tests`.

I'm a bit confused about this, there is no make target for running unit tests any more? Should I use `go test` to run just the unit tests, no make target for it is needed any more?

File: daemon/Makefile
Comment (jrajahalme) [inline daemon/Makefile]: Any reason this would not work on `arm64` as well?

**Comment by jrajahalme** [comment]
> The `go` tool does not run any extra build steps by default, so if we don't commit the generated file, anyone checking out the code and running `go build` or `go test` will get an error about `DatapathSHA256` being undefined.

For the `go build` result to be useful `github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA` would be needed as well. Currently this is generated by `Makefile.defs`.

**Review by nebril — APPROVED** [review]

**Comment by twpayne** [comment]
> > Done. I've renamed the `unit-tests` target to `integration-tests`.
> 
> I'm a bit confused about this, there is no make target for running unit tests any more? Should I use `go test` to run just the unit tests, no make target for it is needed any more?

Yes. I've removed the `unit-tests` target as it would be equivalent to `go test ./...` and by removing the `unit-tests` target completely we'll catch places where `unit-tests` is still used. Eventually we can add back a `unit-tests` target if needed.

> For the `go build` result to be useful `github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA` would be needed as well. Currently this is generated by `Makefile.defs` from `images/cilium/Dockerfile`.

Thanks, done.

**Comment by twpayne** [comment]
test-me-please

File: daemon/Makefile
Comment (joestringer) [reply daemon/Makefile]: Related, how does this not break the arm64 build?

Comment (joestringer) [reply daemon/Makefile]: :fearful: oh, because it's committed to the tree.

If we fail to update this datapath sha file, then it can break upgrades (basically if the datapath sha seems to be the same then there's no datapath changes therefore no need to replace the BPF programs). Let's be super careful here.

File: Makefile
Comment (joestringer) [inline Makefile:196]: Now that the target is renamed, is this `integration-tests` target also running the unit tests? How do we know that both types of tests are continuing to be run as part of the CI? Bear with me, the answer to this might be simple but I couldn't tell from briefly perusing the changes.

What I had in mind was two targets, one for integration tests and one for unit tests, and these targets run the specific type of test you're asking for, and then we make doubly sure that we aren't omitting any of these tests from the testing infrastructure (ie Travis for unprivileged, Jenkins for privileged tests)

File: pkg/envoy/envoy_version_sha256.go
Comment (joestringer) [inline pkg/envoy/envoy_version_sha256.go]: If we're committing these to the tree now, does this cause weirdness with git any time we build? Should we add these files to .gitignore?

Also, do we need to update these SHAs for each release or anything like that? Or are these just dummy values? In which case should we make them clearly dummy values using a magic pattern?

File: pkg/datapath/loader/hash_test.go
Comment (joestringer) [inline pkg/datapath/loader/hash_test.go]: Why is this necessary? Does the test break? The idea behind the "DatapathHash" object is that if you reset it, it goes back to the default SHA so if this test now needs to write the sha inside, then maybe the object isn't resetting correctly (in which case this could indicate a runtime bug).

File: Makefile
Comment (twpayne) [reply Makefile:196]: > Now that the target is renamed, is this `integration-tests` target also running the unit tests?

Yes. `go test ./...` runs all the unit tests. `go test -tags=integration_tests ./...` runs all the unit tests and the integration tests.

If we wanted separate unit and integration test targets we'd have to tag all the unit test files with a build tag so we can exclude them. I don't know if this is worth doing.

File: pkg/envoy/envoy_version_sha256.go
Comment (twpayne) [reply pkg/envoy/envoy_version_sha256.go]: These files should not be in `.gitignore` as they are needed for Cilium to build from a clean checkout.

> Also, do we need to update these SHAs for each release or anything like that?

The SHA256s should be updated automatically whenever you run `make`. As the generated files are checked into git, any changes to them will be visible in `git diff` and should be checked in alongside the commit that causes the SHA256 to change.

File: pkg/datapath/loader/hash_test.go
Comment (twpayne) [reply pkg/datapath/loader/hash_test.go]: This is necessary because [`newDatapathHash()` seeds the hash with `DatapathSHA256`](https://github.com/cilium/cilium/blob/master/pkg/datapath/loader/hash.go#L43).

I'm a bit suspicious of the datapath hashing code, but I consider it a separate issue to making `go test ./...` to succeed by default, so I've done the minimum changes to get the tests to pass in this PR.

**Comment by twpayne** [comment]
test-me-please

File: pkg/envoy/envoy_version_sha256.go
Comment (joestringer) [reply pkg/envoy/envoy_version_sha256.go]: > These files should not be in .gitignore as they are needed for Cilium to build from a clean checkout.

Would adding them to the tree + adding to .gitignore solve this? I understand that we need it available to be able to build, but it sounds like if the build is working correctly then any time we change bpf/ code, this file will change; but we don't really want to track the changes to the file.

> As the generated files are checked into git, any changes to them will be visible in git diff and should be checked in alongside the commit that causes the SHA256 to change.

We used to have a system like this in the past, it's a pain to maintain. This means if there are ever parallel BPF changes in flight, we are forced to rebase BPF PRs purely to satisfy this constraint :/

File: pkg/datapath/loader/hash_test.go
Comment (joestringer) [reply pkg/datapath/loader/hash_test.go]: :thinking: newDatapathHash() always seeded the hash with DatapathSHA256. 

When I look a little closer, it seems like there is no `h.Reset()` outside of unit tests, so the comment on line 45->47 is actually asserting behaviour that is not implemented and not used. That's part of where this discrepancy came from.  Honestly it looks like we should just remove these 4 lines of test + comment.

I think that maybe at some point, `Reset()` was expected to be implemented and used for resumeable hashes (to ensure that the hash was always seeded with the base datapath sha), but in the end we landed with `sumEndpoint()` which makes a copy of the hash every time, so it's actually not necessary.

File: pkg/envoy/envoy_version_sha256.go
Comment (joestringer) [reply pkg/envoy/envoy_version_sha256.go]: ^^ I think we should raise awareness of this proposal for this last point and see if we can come up with a better solution.

**Comment by twpayne** [comment]
test-me-please

**Comment by twpayne** [comment]
test-me-please

**Comment by twpayne** [comment]
test-me-please

**Comment by twpayne** [comment]
`ci-l4lb` test failure is:

```
ERROR: failed to create cluster: failed to pull image "kindest/node:v1.21.1@sha256:fae9a58f17f18f06aeac9772ca8b5ac680ebbed985e266f711d936e91d113bad": command "docker pull kindest/node:v1.21.1@sha256:fae9a58f17f18f06aeac9772ca8b5ac680ebbed985e266f711d936e91d113bad" failed with error: exit status 1

Command Output: Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

**Comment by twpayne** [comment]
ConformanceEKS (ci-eks) failure looks like a flake, I've opened #17102.

**Comment by twpayne** [comment]
k8s-1.19-kernel-5.4 (test-1.19-5.4) failure is #16852.

File: test/Makefile
Comment (joestringer) [inline test/Makefile:28]: Do we need to do this for the `build` target as well?

Upstream docs say:

```
Build flags:
    -tags
        Set -tags to pass in build tags down to the compilation step.
```

File: test/test_test.go
Comment (joestringer) [inline test/test_test.go:4]: It'd be nice if we could figure out a way to allow both `go test ./...` and also `ginkgo -focus ...` without adding this extra `-tags integration_tests` boilerplate argument for every time a developer wants to run ginkgo :thinking: 

At the same time, I personally have macros wrapped around the ginkgo invocations anyway so it doesn't directly affect me. And I'd probably rather that unit-testing is the easiest type of testing in order to encourage developers to write more of them.

**Review by joestringer — APPROVED** [review]
One minor thing to check on the `make -C test build` target. Not sure why you're hitting so much turmoil in the CI, is the tree in general this unstable? :thinking:

File: test/Makefile
Comment (twpayne) [reply test/Makefile:28]: No, we don't need to add `-tags=integration_tests` to the build target as the tag is only used in tests.

File: test/test_test.go
Comment (twpayne) [reply test/test_test.go:4]: > It'd be nice if we could figure out a way to allow both `go test ./...` and also `ginkgo -focus ...` without adding this extra `-tags integration_tests` boilerplate argument for every time a developer wants to run ginkgo 🤔

Agree, I'd prefer this as well, but I can't find a mechanism to do it, short of providing our own `ginkgo` binary.

**Comment by twpayne** [comment]
> Not sure why you're hitting so much turmoil in the CI, is the tree in general this unstable? 🤔

In my personal experience, yes, the tree has been this unstable, but perhaps I've just been unlucky.

File: test/Makefile
Comment (twpayne) [reply test/Makefile:28]: > One minor thing to check on the `make -C test build` target.

Ah, I see what you mean now. AFAICT the resulting `test.test` binary is not used anywhere, but I've updated the target to add the `integration_tests` tag.

**Review by tklauser — APPROVED** [review]

**Review by jrajahalme — APPROVED** [review]

**Comment by twpayne** [comment]
Gilberto is on PTO for a bit, this has multiple approvals, so marking as `ready-to-merge`.

**Review by gandro — APPROVED** [review]
:rocket:

**Pull request merged**
